src/main/scala/org/constellation/domain/redownload/RedownloadService.scala

```diff
@@ -1,38 +1,170 @@
+            s"Snapshot to delete : ${plan.diff.snapshotsToDelete.map(a => (a.height, a.hash))} \n" +
+            s"From peers : ${plan.diff.peers} \n" +
+            s"Own snapshots : ${ownSnaps.values.map(a => (a.height, a.hash)).toList} \n" +
+            s"ReDownloadPlan : ${ReDownloadPlan(plan.localId, plan.sortedSnaps, plan.nodeIdsWithSnaps, plan.diff, plan.groupedProposals)}"
+        ) >>
+          healthChecker
+            .startReDownload(plan.diff, peers.filterKeys(plan.diff.peers.contains))
+            .map(_ => Some(plan.sortedSnaps.toList))
+      } else {
+        Sync[F].pure[Option[List[RecentSnapshot]]](None)
+      }
+    } yield result
+}
 
+case class ReDownloadPlan(localId: Id, sortedSnaps: Seq[RecentSnapshot], nodeIdsWithSnaps: Set[Id], diff: SnapshotDiff,
+                          groupedProposals: Map[Id, Seq[RecentSnapshot]]){
+  def validRedownloadPeers = {
+    nodeIdsWithSnaps
+      .filter{ id =>
+        val peerProposals = groupedProposals.getOrElse(id, Seq()).toSet
+        val gapBetweenHeightsExists = diff.snapshotsToDownload.toSet.diff(peerProposals).nonEmpty
+        !gapBetweenHeightsExists
+      }
+  }
 }
 
 object RedownloadService {
-  def apply[F[_]: Concurrent](): RedownloadService[F] = new RedownloadService[F]()
+  type Proposals = Map[Long, RecentSnapshot]
+
+  val snapshotHeightDelayInterval: Int = ConfigUtil.constellation.getInt("snapshot.snapshotHeightDelayInterval")
+  val fetchSnapshotProposals = "fetchSnapshotProposals"
+  val snapshotHeightRedownloadDelayInterval: Int =
+    ConfigUtil.constellation.getInt("snapshot.snapshotHeightRedownloadDelayInterval")
+
+  private def isMisaligned(ownSnapshots: List[RecentSnapshot], recent: Map[Long, String]) =
+    ownSnapshots.exists(r => recent.get(r.height).exists(_ != r.hash))
+
+  private def isAboveInterval(ownSnapshots: List[RecentSnapshot], snapshotsToDownload: List[RecentSnapshot]) =
+    (maxOrZero(ownSnapshots) - snapshotHeightDelayInterval) > maxOrZero(
+      snapshotsToDownload
+    )
+
+  private def isBelowInterval(ownSnapshots: List[RecentSnapshot], snapshotsToDownload: List[RecentSnapshot]) =
+    (maxOrZero(ownSnapshots) + snapshotHeightDelayInterval) < maxOrZero(
+      snapshotsToDownload
+    )
+
+  def shouldReDownload(ownSnapshots: List[RecentSnapshot], diff: SnapshotDiff, peers: Set[Id]): Boolean = {
+    val needsRedownload = diff match {
+      case SnapshotDiff(_, _, Nil) => false
+      case SnapshotDiff(_, Nil, _) => false
+      case SnapshotDiff(snapshotsToDelete, snapshotsToDownload, _) =>
+        val above = isAboveInterval(ownSnapshots, snapshotsToDownload)
+        val below = isBelowInterval(ownSnapshots, snapshotsToDownload)
+        val misaligned =
+          isMisaligned(ownSnapshots, (snapshotsToDelete ++ snapshotsToDownload).map(r => (r.height, r.hash)).toMap)
+        above || below || misaligned
+    }
+    val canRedownload = peers.nonEmpty//todo log
+    needsRedownload && canRedownload
```

Comment:
You should check for `canRedownload` upfront. There is no point to calculate if node needs redownload if it cannot redownload anyway.